ballerina/modules/wssec/tests/ws_security_tests.bal

```diff
@@ -0,0 +1,1017 @@
+    string envelopeString = securedEnvelope.toString();
+
+    assertUsernameToken(envelopeString, DERIVED_KEY_TEXT);
+}
+
+@test:Config {
+    groups: ["username_token", "password_digest"]
+}
+function testUsernameTokenWithHashedPasword() returns error? {
+    xml envelope =
+    xml `<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
+            <soap:Body>
+                <person>
+                    <name>John Doe</name>
+                    <age>30</age>
+                    <address>
+                        <city>New York</city>
+                        <country>USA</country>
+                    </address>
+                </person>
+            </soap:Body>
+        </soap:Envelope>`;
+    xmlns "http://schemas.xmlsoap.org/soap/envelope/" as soap;
+    UsernameTokenConfig utRecord = {
+        username: USERNAME,
+        password: PASSWORD,
+        passwordType: DIGEST
+    };
+    xml securedEnvelope = check applyUsernameToken(envelope, utRecord);
+    string envelopeString = securedEnvelope.toString();
+
+    assertUsernameToken(envelopeString, DIGEST);
+}
+
+@test:Config {
+    groups: ["username_token", "password_digest", "derived_key"]
+}
+function testUsernameTokenWithHashedPaswordWithDerivedKey() returns error? {
+    xml envelope =
+    xml `<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
+            <soap:Body>
+                <person>
+                    <name>John Doe</name>
+                    <age>30</age>
+                    <address>
+                        <city>New York</city>
+                        <country>USA</country>
+                    </address>
+                </person>
+            </soap:Body>
+        </soap:Envelope>`;
+    xmlns "http://schemas.xmlsoap.org/soap/envelope/" as soap;
+    UsernameTokenConfig utRecord = {
+        username: USERNAME,
+        password: PASSWORD,
+        passwordType: DERIVED_KEY_DIGEST
+    };
+    xml securedEnvelope = check applyUsernameToken(envelope, utRecord);
+    string envelopeString = securedEnvelope.toString();
+
+    assertUsernameToken(envelopeString, DERIVED_KEY_DIGEST);
+}
+
+@test:Config {
+    groups: ["username_token", "signature", "symmetric_binding"]
+}
+function testSymmetricBindingPolicyWithSignatureOnly() returns error? {
+    xml envelope =
+    xml `<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
+            <soap:Body>
+                <person>
+                    <name>John Doe</name>
+                    <age>30</age>
+                    <address>
+                        <city>New York</city>
+                        <country>USA</country>
+                    </address>
+                </person>
+            </soap:Body>
+        </soap:Envelope>`;
+    xmlns "http://schemas.xmlsoap.org/soap/envelope/" as soap;
+
+    crypto:KeyStore serverKeyStore = {
+        path: X509_KEY_STORE_PATH,
+        password: KEY_PASSWORD
```

Comment:
we could define these records as module level readonly records and avoid repeating them in many places.

Reply:
check other places

Reply:
Fixed in [35a7188](https://github.com/ballerina-platform/module-ballerina-soap/pull/103/commits/35a7188066f6d8ccee4b6cd4aa48f754baca1ed4)